### PR TITLE
Make `map2alm` linear again by taking the bad-pixel mask as an argument

### PR DESCRIPTION
**Pull request by ArtemBasyrov**
Fixing [Issue 17](https://github.com/CMBSciPol/jax-healpy/issues/17#issue-4818781994)

Previously, masking would happen dynamically on a map inside `map2alm`, changing the input of SHT functions, hence making them non-linear.  The value-dependent masking was also wrong for iterative use. Applied to a CG residual rather than to the data map, the old code found nothing that looked like `UNSEEN`. It then masked nothing, so the observation mask disappeared mid-solve.

## Change

`map2alm` and `map2alm_spin` take a `bad_mask` argument, broadcastable to the shape of `maps`, and zero exactly those pixels. With the mask supplied from outside, the transform is exactly linear in `maps` for any fixed mask. JAX then transposes it with no custom rule.

`_bad_pixel_mask` is promoted to public `bad_pixel_mask` and exported, so callers can build the mask once.

Example from the issue now works correctly:
```python
mask = jhp.bad_pixel_mask(m)
alm = jhp.map2alm(m, lmax=lmax, bad_mask=mask)

op = lambda v: jhp.map2alm(v, lmax=lmax, bad_mask=mask)
jax.linear_transpose(op, m)(op(m))          # works
```

`smoothing` is unchanged in behavior.It computes the same mask and passes it down.

NB: `map2alm` no longer inspects pixel values. A map containing `UNSEEN` or non-finite values must now be masked explicitly through `bad_mask`. Otherwise those pixels contaminate every coefficient.

## Tests

Added coverage for `linear_transpose`, the adjoint identity `Re<Au, v> == <u, A^T v>`, additivity `f(a + b) == f(a) + f(b)`, mask broadcasting, and `map2alm_spin` with a mask.

### CODE REVIEW

File: jax_healpy/sphtfunc.py [hunk 1]
```diff
@@ -425,16 +426,33 @@ def _alm2map_core(
         return jnp.real(f_complex)
 
 
-def _bad_pixel_mask(m: ArrayLike) -> ArrayLike:
-    """Mask of bad pixels: UNSEEN (via mask_bad) or non-finite (NaN/+-inf)."""
+def bad_pixel_mask(m: ArrayLike) -> ArrayLike:
```
Comment (sbiquard) [inline jax_healpy/sphtfunc.py]: Return type annotation can be more precise I think, `Bool[Array, '...']` in jaxtyping language?

File: jax_healpy/sphtfunc.py [hunk 2]
```diff
@@ -425,16 +426,33 @@ def _alm2map_core(
...
+    --------
+    map2alm : accepts the mask through its ``bad_mask`` argument.
+    """
     m = jnp.asarray(m)
     return mask_bad(m) | ~jnp.isfinite(m)
 
 
-def _sanitize_map(m: ArrayLike) -> ArrayLike:
-    """Zero out bad pixels before a transform, returning a fresh array."""
+def _apply_bad_mask(m: ArrayLike, bad_mask: ArrayLike | None) -> ArrayLike:
```
Comment (sbiquard) [inline jax_healpy/sphtfunc.py]: same comment about return type

File: jax_healpy/sphtfunc.py [hunk 3]
```diff
@@ -829,9 +856,22 @@ def map2alm(
 
     Notes
     -----
-    Bad pixels -- `UNSEEN` or non-finite (``NaN`` / ``+-inf``) -- are replaced by
-    zeros before the transform so they do not contaminate the alm. Input maps are
-    not modified, and each map is masked independently.
+    Like healpy, the transform does not inspect pixel values: a map containing
```
Comment (sbiquard) [inline jax_healpy/sphtfunc.py]: _Unlike_ healpy?

File: jax_healpy/sphtfunc.py [hunk 4]
```diff
@@ -445,6 +463,7 @@ def _map2alm_core(
     iter: int,
     method: str,
     spin: int,
+    bad_mask: ArrayLike | None = None,
 ) -> ArrayLike:
```
Comment (sbiquard) [inline jax_healpy/sphtfunc.py]: here also; but I see that it predates this PR so feel free to defer

File: tests/sphtfunc/test_map_alm.py [hunk 5]
```diff
@@ -401,14 +407,100 @@ def test_map2alm_pol_independent_masks(synthesized_tqu_map: np.ndarray, nside: i
...
 def test_mask_bad_tolerant_and_nan_parity() -> None:
     """mask_bad uses healpy's tolerant comparison and (like healpy) ignores NaN."""
     vals = jnp.array([jhp.UNSEEN, jhp.UNSEEN * (1 + 1e-7), 0.0, np.nan, np.inf])
     mask = jhp.mask_bad(vals)
     np.testing.assert_array_equal(np.asarray(mask), [True, True, False, False, False])
+
+
+def test_bad_pixel_mask_flags_unseen_and_nonfinite() -> None:
+    """bad_pixel_mask extends mask_bad with the non-finite values."""
+    vals = jnp.array([jhp.UNSEEN, jhp.UNSEEN * (1 + 1e-7), 0.0, np.nan, np.inf, -np.inf])
```
Comment (sbiquard) [inline tests/sphtfunc/test_map_alm.py:452]: should there really be a tolerance for how close values are to the sentinel? in the case of `UNSEEN` I would rather use float equality (rtol=atol=0) it is understood as a special value and is not supposed to enter computations so it should always be exactly equal to that number

**Review by sbiquard — DISMISSED** [review]
Thank you! Just a few comments

File: tests/sphtfunc/test_map_alm.py (hunk 5 above)
Comment (ArtemBasyrov) [reply tests/sphtfunc/test_map_alm.py:452]: it's to mirror healpy convention exactly and protect cases, where map went to float32 from float64 or was perturbed somehow otherwise before reading
